File: docker/anki-e2e/artifact-paths.mjs
import fs from "node:fs/promises";
import path from "node:path";

export function resolveArtifactPaths(root = process.env.ANKI_STUDY_REPORT_E2E_ARTIFACTS || "/e2e/artifacts") {
  const reports = path.join(root, "reports");
  const html = path.join(root, "html");
  const screenshots = path.join(root, "screenshots");
  return {
    root,
    runtime: path.join(root, "runtime"),
    diagnostics: path.join(root, "diagnostics"),
    reports,
    html,
    screenshots,
    package: path.join(root, "package"),
    report(fileName) {
      return path.join(reports, fileName);
    },
    htmlFile(...segments) {
      return path.join(html, ...segments);
    },
    pageScreenshot(pageName, theme) {
      return path.join(screenshots, "pages", pageName, `${theme}.png`);
    },
    navigationScreenshot(theme) {
      return path.join(screenshots, "navigation", `avatar-menu-${theme}.png`);
    },
    cardsScreenshot(fixture, mode, theme) {
      return path.join(screenshots, "cards", fixture, modeDirectory(mode), `${theme}.png`);
    },
    failureScreenshot(fileName) {
      return path.join(screenshots, "failures", fileName);
    },
  };
}

export async function ensureArtifactParent(filePath) {
  await fs.mkdir(path.dirname(filePath), { recursive: true });
}

export function relativeArtifactPath(paths, filePath) {
  return path.relative(paths.root, filePath).split(path.sep).join("/");
}

function modeDirectory(mode) {
  return mode === "ankiPreview" ? "anki-preview" : mode;
}

File: docker/anki-e2e/artifact_paths.py
from __future__ import annotations

from dataclasses import dataclass
import os
from pathlib import Path


@dataclass(frozen=True)
class ArtifactPaths:
    root: Path
    runtime: Path
    diagnostics: Path
    reports: Path
    html: Path
    screenshots: Path
    package: Path

    @classmethod
    def from_root(cls, root: str | Path) -> "ArtifactPaths":
        resolved = Path(root)
        return cls(
            root=resolved,
            runtime=resolved / "runtime",
            diagnostics=resolved / "diagnostics",
            reports=resolved / "reports",
            html=resolved / "html",
            screenshots=resolved / "screenshots",
            package=resolved / "package",
        )

    @classmethod
    def from_env(cls) -> "ArtifactPaths":
        return cls.from_root(os.environ.get("ANKI_STUDY_REPORT_E2E_ARTIFACTS", "/e2e/artifacts"))

    def ensure(self) -> None:
        for directory in (
            self.root,
            self.runtime,
            self.diagnostics,
            self.reports,
            self.html,
            self.screenshots,
            self.package,
        ):
            directory.mkdir(parents=True, exist_ok=True)

    def relative(self, path: str | Path) -> str:
        return Path(path).relative_to(self.root).as_posix()

